### PR TITLE
Structured Generation for the Simulator

### PR DESCRIPTION
**Pull request by alpaylan**
- added Arbitrary and ArbitraryOf<T> traits for mroe centralized generation
- implemented random generation for tables and structured queries

### CODE REVIEW

File: simulator/main.rs
Comment (jussisaurio) [inline simulator/main.rs:44]: I feel like `ArbitraryFrom<T>` and `arbitrary_from()` are more accurate names for these?

take e.g.

```
// Arbitrary Value from Vec<&Value>
impl ArbitraryFrom<Vec<&Value>> for Value {
    fn arbitrary_from<R: Rng>(rng: &mut R, t: &Vec<&Value>) -> Self {
        if t.is_empty() {
            return Value::Null;
        }

        let index = rng.gen_range(0..t.len());
        t[index].clone()
    }
}
```

or

```
// Arbitrary Value from Columntype
impl ArbitraryFrom<ColumnType> for Value {
    fn arbitrary_from<R: Rng>(rng: &mut R, t: &ColumnType) -> Self {
        match t {
            ColumnType::Integer => Value::Integer(rng.gen_range(i64::MIN..i64::MAX)),
            ColumnType::Float => Value::Float(rng.gen_range(-1e10..1e10)),
            ColumnType::Text => Value::Text(gen_random_text(rng)),
            ColumnType::Blob => Value::Blob(gen_random_text(rng).as_bytes().to_vec()),
        }
    }
}
```

File: simulator/main.rs
Comment (jussisaurio) [inline simulator/main.rs:243]: IMO either call this `where` or `predicate`, since we already use a `Predicate` type here (and the `predicate` name elsewhere in Limbo too, actually)

**Comment by jussisaurio** [comment]
IMO this looks like a great direction 👍

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs:44]: Agreed, I had the same thought , but I had already written most of the code so just kept it as is at that point. I'll update them.

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs:246]: We can't call it where, as Rust already has `where` as a keyword. I changed it to `predicate`.

**Comment by alpaylan** [comment]
> IMO this looks like a great direction 👍

Glad to hear that! I feel like after https://github.com/tursodatabase/limbo/pull/469 is merged, I can change the current workaround to be able to run the simulator (I had changed `write!(f, "TRUE")` to `write!(f, "1 = 1")`), and we can merge this in its current form and gradually improve. What do you think @jussisaurio ?

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:621]: So now we dont have workload types. What I like of what I was doing is that there would be a Read/write/delete distribution where all the queries would derive from.

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:258]: I believe this would be nicer if it followed the workload distribution. What do you think

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:182]: What does gt mean

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:141]: What does lt mean

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:155]: Nit s/t/value

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:173]: What is happening here

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:191]: I dont know what gtvalue so idk what to expect from this

**Review by pereman2 — COMMENTED** [review]
I left some commentd because I dont understand the terminology of anything and it basically removes the idea of workloads from the simulator.

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs:258]: I think I missed the workload distribution part when I was working through the code. I'll roll back and try to understand how that fit into the previous scheme and bring it back.

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs:182]: The GT and LT in this context mean `GreaterThan` and `LessThan`. They're not real types, it's just newtype pattern because of the trait oriented generation mechanism. We can also switch to using explicit functions which could reduce the mental overhead.

A `GTValue` represents a `Value` `v2` that satisfies the `>` predicate with respect to another value `v1`. `GTValue` and `LTValue` don't have `Arbitrary` instances, but only `ArbitraryFrom<Value>`, in which they generate values that are either greater than, or less than the provided value.

Does that make sense?

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs:141]: Please see the explanation for the previous comment.

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs:155]: Makes sense 👍🏻

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs:173]: There are two ways of generating a smaller text with respect to the lexicographic ordering:

1- `abc` is "smaller" than `abcd`, in the sense that if two strings share the same prefix, shorter is less than the longer. In 1% of the cases, the code removes the last element of the given value to create a shorter one.

2- `abc` is "smaller" than `aac`, or `aaz`, as the lexicographic ordering is the comparison of two char arrays starting with the leftmost character. The code first picks an index to `make smaller`, as in converting `abc` to `aac` by mutating `b` to `a`, and than randomly generating the right side of the picked index, as it is not important anymore.

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:311]: please try to make variables more meaningful I don't know what b is here.

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:549]:
```suggestion
    let table = &env.tables[table];

    // Pick a random column
    let column_index = env.rng.gen_range(0..table.columns.len());
    let column = &table.columns[column_index].clone();
```

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:574]: What if the row is not inserted in order? The idea was to keep the rows in the same order as the table would to improve lookup logic.

File: simulator/main.rs
Comment (pereman2) [reply simulator/main.rs:182]: To be honest what I'm finding hard to wrap around in this PR is the move from simple functions to a whole bunch of abstract indirection that I'm still trying to parse. I don't say it's bad not at all. I'm just saying that it will take me some time to ensure this is what we want. Maybe you could provide me with some literature with regards to this kind of testing? I want to be sure I understand everything before merging because it seems interesting but I just don't want to merge something based on coolness. I also feel like we need more comments because there are a lot of abbreviations that make it hard to follow.

**Review by pereman2 — CHANGES_REQUESTED** [review]

File: simulator/main.rs
Comment (jussisaurio) [reply simulator/main.rs:311]: yeah i agree with this, let's use longer more descriptive variable names

File: simulator/main.rs
Comment (jussisaurio) [reply simulator/main.rs:182]: agreed w/ more comments

Comment (alpaylan) [reply simulator/main.rs:182]: Yeah, I fully understand and agree with that. Let me try to address your points a bit.

- The indirections themselves might actually be useless for now, I've been using them because that's the style of writing I'm used to, so if they're creating extra mental load, we can drop them.
- The testing style itself comes from [QuickCheck](https://hackage.haskell.org/package/QuickCheck), the Property-Based Testing library from Haskell. The trait based generators are beneficial for implementing (1) higher order generator functions such as a generator that picks one of a few given options. You might've realized we use `rng.gen_range()` to do weighted random selection, QuickCheck has a `frequency` function that does that in a safer way, leveraing the typeclasses. The second benefit of the traits are that you can automatically derive properties from functions, given that the inputs themselves have generators defined for them(in other words, they implement `Arbitrary` trait). Here's a [Rust implementation of QuickCheck](https://github.com/BurntSushi/quickcheck)
- I also fully agree that we shouldn't add anything just because it looks fancy, I wouldn't have written the PR if I thought the main benefit was that. Let me state the contributions I see here:

1. I personally think the structured approach(creating a query type, randomly picking queries) scales better than the ad hoc approach in the current version. I have plans of adding more complex query types within the next week such as `Select .. Limit` or `OrderBy`.
2. It adds the `where` clauses to Select and Delete queries, and randomly generates those clauses. 

In my opinion, much of the complexity of the code comes from the second part. Generation of such queries is (I think) inherently complex, because generating an expression that satisfies a constraint is complex. I agree that they could be better documented, both by inline comments and a separate documentation discussing the generation strategies and properties.

What do you think? What should the forward steps be?

Comment (jussisaurio) [reply simulator/main.rs:182]: Yeah a lot of this IMO hinges on whether we want the creation of arbitrary queries to be part of the simulator. If yes, then I see this kind of property-testing setup as a big plus cuz instead of having a bunch of if-else soup you can have a `Query::arbitrary()` type thing and encapsulate whatever "arbitrary" means inside those trait implementations. One big reason this PR has more complexity than `main` is that it's just doing more complex things than `main` is, as you said.

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs]: I've added the workloads as one of the modes, let me know how that sounds, or if you think it's better to keep the workloads as the main testing system.

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs]: Changed all the one character names to longer and descriptive versions.

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs]: I've changed all the names to more descriptive versions.

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs]: This change cases a problem with the borrow checker, at least as far as I was able to work with it.

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs:670]: Should the push be randomized?

Comment (pereman2) [reply simulator/main.rs:670]: No, I just mean the env.tables[table].rows should probably be ordered the same way the rows would be ordered in disk. This way if we want to compare the rows we can do it quickly.

File: simulator/main.rs
Comment (pereman2) [reply simulator/main.rs]: Hmmm I don't think we want different modes. I think what we want is to randomize the environment first and then perform the simulation based on the parameters of the environment. For example at startup we decide workload distribution, maximum connections, number of tables we will want, how big we want texts to be, % of IO faults, etc... A different mode would mean that we fix the environment to something instead of having explicit logic if..else.

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:295]:
```suggestion
impl Arbitrary<Vec<Table>> for Select {
```
nit

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:296]:
```suggestion
    fn gen<R: Rng>(rng: &mut R, tables: &Vec<Table>) -> Self {
```
Nit? What do you think

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:482]: We need to also test where `WHERE` does not exist too right?

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:473]: what does this boolean do?

File: simulator/main.rs
Comment (pereman2) [inline simulator/main.rs:410]: These random constants would nicer to have extracted somewhere else to understand why chose a specific percentage. These are things where a workload config is nice to be setup from the start, because we can change all of these constants to something else to create a completely different scenario.

File: simulator/main.rs
Comment (pereman2) [reply simulator/main.rs:205]: I'm getting around these properties let's see how it goes but let's first try to clean up some things. If you can leave comments and improve naming of things that would help.

**Review by pereman2 — CHANGES_REQUESTED** [review]

File: simulator/main.rs
Comment (jussisaurio) [reply simulator/main.rs:295]: that's a different thing. `impl Arbitrary for MyType` is `"make an arbitrary value of type MyType"`, while `impl ArbitraryFrom<SourceType> for MyType` is "Make an arbitrary value of type MyType from SourceType", for example:

```
impl Arbitrary for Table {
  /* make an arbitrary table with arbitrary columns of arbitrary types */
}
```

```
impl ArbitraryFrom<Table> for Select {
  /* Make an arbitrary Select statement that reads from Table */
}
```

File: simulator/main.rs
Comment (jussisaurio) [reply simulator/main.rs:473]: yeah please name it better

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs:482]: That makes sense, I think we ideally would like this type to reach the whole of `Select` type in the limbo AST. 

```rust
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Select {
    /// CTE
    pub with: Option<With>,
    /// body
    pub body: SelectBody,
    /// `ORDER BY`
    pub order_by: Option<Vec<SortedColumn>>, // ORDER BY term does not match any column in the result set
    /// `LIMIT`
    pub limit: Option<Limit>,
}

/// `SELECT` body
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct SelectBody {
    /// first select
    pub select: OneSelect,
    /// compounds
    pub compounds: Option<Vec<CompoundSelect>>,
}
...
#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CompoundSelect {
    /// operator
    pub operator: CompoundOperator,
    /// select
    pub select: OneSelect,
}

/// Compound operators
// https://sqlite.org/syntax/compound-operator.html
#[derive(Copy, Clone, Debug, PartialEq, Eq)]
pub enum CompoundOperator {
    /// `UNION`
    Union,
    /// `UNION ALL`
    UnionAll,
    /// `EXCEPT`
    Except,
    /// `INTERSECT`
    Intersect,
}

/// `SELECT` core
// https://sqlite.org/syntax/select-core.html
#[derive(Clone, Debug, PartialEq, Eq)]
pub enum OneSelect {
    /// `SELECT`
    Select {
        /// `DISTINCT`
        distinctness: Option<Distinctness>,
        /// columns
        columns: Vec<ResultColumn>,
        /// `FROM` clause
        from: Option<FromClause>,
        /// `WHERE` clause
        where_clause: Option<Expr>,
        /// `GROUP BY`
        group_by: Option<GroupBy>,
        /// `WINDOW` definition
        window_clause: Option<Vec<WindowDef>>,
    },
    /// `VALUES`
    Values(Vec<Vec<Expr>>),
}
```

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs:473]: Changed to `predicate_value`

File: simulator/main.rs
Comment (alpaylan) [reply simulator/main.rs:410]: Yeah agreed. This specific one is rather arbitrary, I can just turn it into an equal probability version.

**Comment by alpaylan** [comment]
@pereman2 @jussisaurio I wanted to point out several design decisions worth discussing:

The current mode of generation is online, in other words, we decide on the next query as we execute the current one. This is one potential approach, which I think does not always benefit us. The reason I feel this is related to the idea of properties.

In the previous version of the simulator, correctness was based on modeling, or shadowing. We have a model of each operation, and we keep a replica of the database with respect to those models. Whenever we create an insert query, we push the same value to the replica table we have, whenever we run a select query, we check the result rows are those that exist in the replicate table.

I personally think this approach does not scale well to testing the entirety of limbo. Keeping the model up with limbo functionality is not going to be easy as the database gets more complex, because we'll have to reimplement every functionality we want to test in the database in the simulator model.

The approach I had taken was to define properties on the database, without necessarily using a separate model. `property_insert_select` is a good example. The correctness does not depend on a model, but the expectations from the database itself.

`property_insert_select` tells us that running `INSERT` on a value `V`, then running a`SELECT` query that has a where clause that evaluate to `TRUE` for the value `V` should give us at least one row, and the resulting row set should include `V`. We can test this property without keeping a replicate database ourselves.

Other examples of such properties would be;

- Adding an `AND TRUE` to a where clause should not change the output of a select query
- Adding an `OR FALSE` to a where clause should not change the output of a select query
- Running the same `SELECT` query two times, without any `INSERT` statements in the middle, should yield the same output
- Running select `V` for a value should not contain the value is it was deleted and not added beforehand.

Of course, my initial property-based approach was naive, in the sense that it solely focused on the correctness of query execution, not thinking of database as a live object, so I didn't understand why the workloads was an important part. That's why I think an **offline** generation approach might be better now.

In the **offline generation**, we first generate an array of `queries and assertions` to run on the database. This part again requires us to have some model of the database as we want to create meaningful queries. The generation would have several constraints:

1. We generate queries based on properties we want to test at certain points. The `property_insert_select` would generate an array of 3 items (2 queries and 1 assertion). `INSERT V ; SELECT V ; assert V in RESULT`.
2. We generate queries respecting the overall expected workload. Properties have flexibilities, in which you can actually insert statements between them. Going back to `insert_select`, we can inject as many `INSERT (NOT V)` or `SELECT ANY` statements in the middle without breaking the property, so as long as we don't insert another `V` or somehow delete or update the existing `V`. Leveraging this, we can dynamic adjust generation probabilities(because we have access to the overall list of statements to execute`.
3. Currently, it's sometimes possible to generate invalid statements, crashing the simulator. If we create a table with the same column name twice, the parser fails, and we currently sometimes create such tables. It makes sense to have a test where that happens, but I don't think that's what the simulator is testing. Using the offline generation, we can just discard such generations that don't conform to our assertions.

The harder part about this is that we need to come up with those properties I mentioned. Some of those queries already exist in database testing literature, so it's easy to borrow.

This approach has two more nice things about it:

1. It's very easy to produce a corpus of failing query arrays, making it easier to handle regressions by running previously failing query arrays. Perhaps this becomes kind of irrelevant with DST though, because the seed keeps all of this information already at hand.
2. We can now run `shrinking` on the results by systematically making the query array smaller, automatically receiving minimum reproducible examples, compared to the online generation where debugging will certainly become a mess for complex bugs.

Please let me know what you think about it so we can move forward.

**Comment by pereman2** [comment]
> I personally think this approach does not scale well to testing the entirety of limbo. Keeping the model up with limbo functionality is not going to be easy as the database gets more complex, because we'll have to reimplement every functionality we want to test in the database in the simulator model.

Maybe we could add a `strict` mode where we test the output of a query with sqlite3. But maybe this is a bad idea because sqlite3 could have the same bug.

With regards to online vs offline, I don't mind if its done first or while the process, at the end the test should be reproducible through the seed right?. In offline runs I see the benefit of `dry-run` of seeds we know fail. 

Anyways, I loved your explanation. I agree it seems like it will not scale well because if we have to reimplement every query logic again it would be bad. I would say I would still love to see the environment setup of percentages of stuff at the beginning and for the rest it seems great.

**Comment by alpaylan** [comment]
> Maybe we could add a `strict` mode where we test the output of a query with sqlite3. But maybe this is a bad idea because sqlite3 could have the same bug.

I actually like this idea a lot, it's the simplest property we could test. We should definitely have it as one of the options.

> With regards to online vs offline, I don't mind if its done first or while the process, at the end the test should be reproducible through the seed right?. In offline runs I see the benefit of `dry-run` of seeds we know fail.

Yeah I think it would be reproducible in either case, but I still think having a corpus is nice. It's also nice in the sense that you can directly give it to a fuzzer and ask it to test things. So it opens up other possibilities too.

> Anyways, I loved your explanation. I agree it seems like it will not scale well because if we have to reimplement every query logic again it would be bad. I would say I would still love to see the environment setup of percentages of stuff at the beginning and for the rest it seems great.

Yeah I agree with the percentages. I just broke down everything into parts, and today I'll work on getting it into a better shape while increasing the documentation coverage.

**Comment by alpaylan** [comment]
Hi @pereman2,

This latest change adds the offline generation via interaction plans, found in `simulator/generation/plan.rs`. A plan is basically a list of queries and assertions, where queries are our typical sql queries, and the assertions are Rust closures reasoning about the results of the queries themselves.

The query plans respect the workload distributions, handled via the `frequency` calculation at `simulator/generation/plan.rs/L322`. You can see an example of how to write a property with `simulator/generation/plan.rs/property_insert_select` at line 246. 

When an assertion fails, we write the query plan at a temporary directory the same as the database as a valid set of sql statements, so that it's easy to use and reproduce. The seeds themselves are(at least to me) enough for reproducing too, as I've been using them to reproduce failing experiments.

What's missing?

- [ ] There are no fault injections for now. We can add fault injections as part of the query plans. As I didn't fully know how they worked, I didn't wanna touch them.
- [ ] No shrinking: Ideally, when a test fails, we should be able to obtain a smaller test that fails too. I plan to implement that as one of the next steps.
- [ ] Complex generation: For now, I only added the `eq` check in the `property_insert_select`. The complex generation strategies implemented in `simulator/generation/query.rs` will give a better test coverage in the long term.
- [ ] CLI: Another long term feature I wanna have is a CLI in front of the simulator to run configuration independent of the code, which we currently do in ad-hoc manner using the environment variables such as SEED.
- [ ] Documentation: The current documentation is not very comprehensive. Please let me know if you see any undocumented code that is cryptic.

The discussion related to the code itself has been quite a mess in the previous comments, so I suggest that the future discussion moves forward independent of them, maybe by resolving all of them even we open duplicates after this comment.  I would love to resolve them in a linear fashion, as the PR itself is quite big at this point, causing even more confusion as time goes on.

**Review by pereman2 — APPROVED** [review]
Looking great! I like the plan file.

Frew things:
1. I would love to the see interaction printed with info logs. Mainly to be able to trace back which interaction number corresponds with the plan. 
2. Looks like tables are not being created enough. Before, write_percent also encompassed `CREATE table` because it is inherently a write to disk, I want to propose a hierarchical structure to some type of writes/reads, because you would probably want something like 34% writes and 1% of those writes to be create tables. I say this because I have tons of bugs to solve with CREATE table and I was able to see them before but hard to see them with this implementation.

I think there is no need to do this here and I want to unblock you so we can fix in later PRs.

**Comment by pereman2** [comment]
btw there are CI failures.

**Comment by alpaylan** [comment]
@pereman2 added the interaction logs, increased the table creation probability at `generation/plan.rs`, and fixed the build problems.

**Pull request closed without merging**
